### ISSUE TITLE
Restricted users cannot use "save and new" in the news module

### ISSUE THREAD

**Issue by monxoom** [contao/contao#2201]
**Affected version(s)**

Contao 4.8+ (maybe older ones, too)

**Description**

Create a news item and then a CE 'text'. Write some text and then use the editor's 'insert link' feature to link to another news item. Try to save the CE with the button "Save and new" – works (if logged in as admin). Try the same but as restricted user – and it fails. 

What I've found out: 
- It fails because checkAccessToElement() fails (in class 'tl_content_news')
- the method fails because the methods first SQL query yields an empty result 
- result is empty because $id is strange
- $id resolves to CURRENT_ID
- CURRENT_ID gets filled with $objSession->get('CURRENT_ID')
- the session's CURRENT_ID gets changed as soon as you use the 'insert link' feature and choose a news archive (to choose another news item)

Resulting error (taken from the Contao demo installation):
![Bildschirmfoto 2020-08-26 um 18 42 13](https://user-images.githubusercontent.com/1366842/91333415-dbca4680-e7cd-11ea-9a9e-327360177f68.png)

The content element doesn't have the ID 1, but the news archive does have this ID …

Tested in a Contao 4.8.8 installation and in the Contao demo installation.

**Comment by leofeyer** [comment]
@aschempp @fritzmg Can we solve this by somehow passing the current ID to the picker context?

**Comment by aschempp** [comment]
we already do similar with `source`/`preserveRecord` for pages? https://github.com/contao/contao/blob/master/core-bundle/src/Picker/PagePickerProvider.php#L71

**Comment by monxoom** [comment]
*push*

**Comment by leofeyer** [comment]
The issue is not actually caused by the picker. We are overwriting `CURRENT_ID` here as soon as the picker is opened:

https://github.com/contao/contao/blob/214825ce58eef8174c353ee094f1e4f7df85bf71/core-bundle/src/Resources/contao/classes/Backend.php#L343-L347

The only fix I can think of is to use two different `CURRENT_ID` values, one for the regular back end and one for the picker/popup back end. @contao/developers WDYT?

**Comment by leofeyer** [comment]
As discussed in Mumble on August 26th, @ausi will try to remove `CURRENT_ID` from the `tl_content_news::checkPermissions()` method.

**Comment by ausi** [comment]
~~I debugged the problem and it seems that the error occurs because of `reviseTable()`.
Without the following line the error disappears:~~ https://github.com/contao/contao/blob/0ba1acd7d93da55598a02fa6963dacf635bdad87/core-bundle/src/Resources/contao/drivers/DC_Table.php#L3374 ~~@monxoom can you verify that the error is fixed if you delete that line?~~

~~`tl_content_news::checkPermissions()` is not a problem in this case because it just checks for `act=edit` if the ID of the content element is part of an allowed news archive.~~

~~I was able to reproduce the error with an admin user too, in this case the `Cannot load record "tl_content.id=123".` is shown.~~

~~Exact steps to reproduce in Contao 4.9:~~
1. ~~Edit a news article~~
2. ~~Create a new content element inside it~~
3. ~~Click on insert-link in the tinyMCE editor~~
4. ~~Open the picker~~
5. ~~Go to any news article and click on edit (so that the content elements are shown)~~
6. ~~Cancel the picker~~
7. ~~Click save.~~
8. ~~`Cannot load record "tl_content.id=123".`~~

Created a new issue for this problem: #3597

**Comment by leofeyer** [comment]
@monxoom can you verify that the error is fixed if you delete that line?

**Comment by monxoom** [comment]
I’m sorry, but that doesn’t do the trick for me. Deleted the line in a Contao 4.9 installation and cleared the cache, but the behaviour is exactly as before.

But for the second scenario ausi describes, out-commenting the mentioned line fixes the error!

**Comment by ausi** [comment]
@monxoom I was now able to reproduce your issue.

I think the following changes should fix this error:

```diff
--- a/news-bundle/src/Resources/contao/dca/tl_content.php
+++ b/news-bundle/src/Resources/contao/dca/tl_content.php
@@ -59,12 +59,16 @@ class tl_content_news extends Contao\Backend
                {
                        case '': // empty
                        case 'paste':
-                       case 'create':
                        case 'select':
                                // Check access to the news item
                                $this->checkAccessToElement(CURRENT_ID, $root, true);
                                break;

+                       case 'create':
+                               // Check access to the parent element if a content element is created
+                               $this->checkAccessToElement(Contao\Input::get('pid'), $root, (Contao\Input::get('mode') == 2));
+                               break;
+
                        case 'editAll':
                        case 'deleteAll':
                        case 'overrideAll':
```

Can you confirm that this works?

**Comment by monxoom** [comment]
No luck, sorry! (ofc cleared cache etc.)

**Comment by ausi** [comment]
which specific contao version are you using?
can you post your patched *news-bundle/src/Resources/contao/dca/tl_content.php* file?

**Comment by monxoom** [comment]
I was using the 4.9 LTS (v4.9.22). Here’s the whole patched file:

```php
<?php

/*
 * This file is part of Contao.
 *
 * (c) Leo Feyer
 *
 * @license LGPL-3.0-or-later
 */

// Dynamically add the permission check and parent table
if (Contao\Input::get('do') == 'news')
{
	$GLOBALS['TL_DCA']['tl_content']['config']['ptable'] = 'tl_news';
	array_unshift($GLOBALS['TL_DCA']['tl_content']['config']['onload_callback'], array('tl_content_news', 'checkPermission'));
	$GLOBALS['TL_DCA']['tl_content']['config']['onload_callback'][] = array('tl_content_news', 'generateFeed');
}

/**
 * Provide miscellaneous methods that are used by the data configuration array.
 *
 * @property Contao\News $News
 *
 * @author Leo Feyer <https://github.com/leofeyer>
 */
class tl_content_news extends Contao\Backend
{
	/**
	 * Import the back end user object
	 */
	public function __construct()
	{
		parent::__construct();
		$this->import('Contao\BackendUser', 'User');
	}

	/**
	 * Check permissions to edit table tl_content
	 */
	public function checkPermission()
	{
		if ($this->User->isAdmin)
		{
			return;
		}

		// Set the root IDs
		if (empty($this->User->news) || !is_array($this->User->news))
		{
			$root = array(0);
		}
		else
		{
			$root = $this->User->news;
		}

		// Check the current action
		switch (Contao\Input::get('act'))
		{
			case '': // empty
			case 'paste':
			case 'select':
				// Check access to the news item
				$this->checkAccessToElement(CURRENT_ID, $root, true);
				break;

			case 'create':
				// Check access to the parent element if a content element is created
				$this->checkAccessToElement(Contao\Input::get('pid'), $root, (Contao\Input::get('mode') == 2));
				break;

			case 'editAll':
			case 'deleteAll':
			case 'overrideAll':
			case 'cutAll':
			case 'copyAll':
				// Check access to the parent element if a content element is moved
				if (in_array(Contao\Input::get('act'), array('cutAll', 'copyAll')))
				{
					$this->checkAccessToElement(Contao\Input::get('pid'), $root, (Contao\Input::get('mode') == 2));
				}

				$objCes = $this->Database->prepare("SELECT id FROM tl_content WHERE ptable='tl_news' AND pid=?")
										 ->execute(CURRENT_ID);

				/** @var Symfony\Component\HttpFoundation\Session\SessionInterface $objSession */
				$objSession = Contao\System::getContainer()->get('session');

				$session = $objSession->all();
				$session['CURRENT']['IDS'] = array_intersect((array) $session['CURRENT']['IDS'], $objCes->fetchEach('id'));
				$objSession->replace($session);
				break;

			case 'cut':
			case 'copy':
				// Check access to the parent element if a content element is moved
				$this->checkAccessToElement(Contao\Input::get('pid'), $root, (Contao\Input::get('mode') == 2));
				// no break

			default:
				// Check access to the content element
				$this->checkAccessToElement(Contao\Input::get('id'), $root);
				break;
		}
	}

	/**
	 * Check access to a particular content element
	 *
	 * @param integer $id
	 * @param array   $root
	 * @param boolean $blnIsPid
	 *
	 * @throws Contao\CoreBundle\Exception\AccessDeniedException
	 */
	protected function checkAccessToElement($id, $root, $blnIsPid=false)
	{
		if ($blnIsPid)
		{
			$objArchive = $this->Database->prepare("SELECT a.id, n.id AS nid FROM tl_news n, tl_news_archive a WHERE n.id=? AND n.pid=a.id")
										 ->limit(1)
										 ->execute($id);
		}
		else
		{
			$objArchive = $this->Database->prepare("SELECT a.id, n.id AS nid FROM tl_content c, tl_news n, tl_news_archive a WHERE c.id=? AND c.pid=n.id AND n.pid=a.id")
										 ->limit(1)
										 ->execute($id);
		}

		// Invalid ID
		if ($objArchive->numRows < 1)
		{
			throw new Contao\CoreBundle\Exception\AccessDeniedException('Invalid news content element ID ' . $id . '.');
		}

		// The news archive is not mounted
		if (!in_array($objArchive->id, $root))
		{
			throw new Contao\CoreBundle\Exception\AccessDeniedException('Not enough permissions to modify article ID ' . $objArchive->nid . ' in news archive ID ' . $objArchive->id . '.');
		}
	}

	/**
	 * Check for modified news feeds and update the XML files if necessary
	 */
	public function generateFeed()
	{
		/** @var Symfony\Component\HttpFoundation\Session\SessionInterface $objSession */
		$objSession = Contao\System::getContainer()->get('session');

		$session = $objSession->get('news_feed_updater');

		if (empty($session) || !is_array($session))
		{
			return;
		}

		$request = Contao\System::getContainer()->get('request_stack')->getCurrentRequest();

		if ($request)
		{
			$origScope = $request->attributes->get('_scope');
			$request->attributes->set('_scope', 'frontend');
		}

		$this->import('Contao\News', 'News');

		foreach ($session as $id)
		{
			$this->News->generateFeedsByArchive($id);
		}

		$this->import('Contao\Automator', 'Automator');
		$this->Automator->generateSitemap();

		if ($request)
		{
			$request->attributes->set('_scope', $origScope);
		}

		$objSession->set('news_feed_updater', null);
	}
}
```

**Comment by ausi** [comment]
And the error message you get is still exactly `Invalid news content element ID 1.`?

Please also post the URL that is shown in the browser address bar when you see this error message.

**Comment by monxoom** [comment]
Your questions got me asking myself if … and yes, in a fresh 4.9.22 installation the error doesn’t appear (sorry for taking your time)! On friday I tested with an up-to-date but long running installation where the customer needs this functionality.
I have to recreate the bug in a fresh installation and check back when having more info.

**Comment by ausi** [comment]
I think we need to implement the fix from https://github.com/contao/contao/issues/2201#issuecomment-950293871

**Comment by ausi** [comment]
Closed in favor of #4299
